File: next/plugins/fileservice/fileservice.py
"""C3: viewer.file-service — resolve/read/hash (framework §10.2).

RPC surface: `file:_:resolve` (stat + sha256), `file:_:read` (inline content
up to a byte cap, utf-8 or base64), `file:_:hash` (sha256 only). Large-file
by-reference reads are the http-gateway data plane and out of scope here.

TODO(§16-5 unresolved): path access is currently unrestricted beyond `~`
expansion + absolutization; the allowlist/sandbox policy is an open design
question and lands here once decided.
"""

from __future__ import annotations

import argparse
import asyncio
import base64
import hashlib
import logging
import os
import signal
from pathlib import Path
from typing import Any

from sdk import Ctx, Plugin, slot

logger = logging.getLogger(__name__)

MANIFEST: dict[str, Any] = {
    "id": "file-service",
    "version": "0.1.0",
    "slots": {"resolve": {}, "read": {}, "hash": {}},
    "emits": {},
}

DEFAULT_MAX_READ_BYTES = 1024 * 1024
_HASH_CHUNK = 64 * 1024


class FileServicePlugin(Plugin):
    manifest = MANIFEST

    # ------------------------------------------------------------------ slots

    @slot("file:_:resolve")
    async def resolve(self, ctx: Ctx) -> None:
        path = _path(ctx)
        if path is None:
            await ctx.respond_error("invalid_request", "missing required field: path")
            return
        try:
            stat = path.stat()
        except FileNotFoundError:
            await ctx.respond_error("not_found", f"no such file: {path}")
            return
        except OSError as exc:
            await ctx.respond_error("read_error", str(exc))
            return
        await ctx.respond(
            {
                "path": str(path),
                "exists": True,
                "is_dir": path.is_dir(),
                "size": stat.st_size,
                "mtime": int(stat.st_mtime),
                "sha256": await asyncio.to_thread(_sha256, path) if path.is_file() else None,
            }
        )

    @slot("file:_:read")
    async def read(self, ctx: Ctx) -> None:
        path = _path(ctx)
        if path is None:
            await ctx.respond_error("invalid_request", "missing required field: path")
            return
        max_bytes = int(ctx.value.get("max_bytes", DEFAULT_MAX_READ_BYTES))
        try:
            size = path.stat().st_size
        except FileNotFoundError:
            await ctx.respond_error("not_found", f"no such file: {path}")
            return
        except OSError as exc:
            await ctx.respond_error("read_error", str(exc))
            return
        if size > max_bytes:
            await ctx.respond_error(
                "too_large",
                f"{path} is {size} bytes, above the {max_bytes}-byte inline cap; "
                "use the gateway by-reference data plane",
            )
            return
        try:
            raw = await asyncio.to_thread(path.read_bytes)
        except OSError as exc:
            await ctx.respond_error("read_error", str(exc))
            return
        try:
            content: dict[str, Any] = {"encoding": "utf-8", "content": raw.decode("utf-8")}
        except UnicodeDecodeError:
            content = {
                "encoding": "base64",
                "content": base64.b64encode(raw).decode("ascii"),
            }
        await ctx.respond({"path": str(path), "size": size, **content})

    @slot("file:_:hash")
    async def hash_(self, ctx: Ctx) -> None:
        path = _path(ctx)
        if path is None:
            await ctx.respond_error("invalid_request", "missing required field: path")
            return
        if not path.is_file():
            await ctx.respond_error("not_found", f"no such file: {path}")
            return
        await ctx.respond({"path": str(path), "sha256": await asyncio.to_thread(_sha256, path)})

    # ------------------------------------------------------------- entrypoint

    def run(self) -> None:
        parser = argparse.ArgumentParser(description="Viewer C3 file-service plugin")
        parser.add_argument("--kernel-ws", required=True)
        args = parser.parse_args()
        logging.basicConfig(
            level=logging.INFO, format="%(asctime)s %(levelname)s %(name)s: %(message)s"
        )
        managed = os.environ.get("VIEWER_MANAGED") == "1"

        async def main() -> None:
            await self.start(args.kernel_ws, managed=managed)
            stopped = asyncio.Event()
            loop = asyncio.get_running_loop()
            for signum in (signal.SIGINT, signal.SIGTERM):
                loop.add_signal_handler(signum, stopped.set)
            await stopped.wait()
            await self.stop()

        asyncio.run(main())


def _path(ctx: Ctx) -> Path | None:
    if not isinstance(ctx.value, dict) or not ctx.value.get("path"):
        return None
    return Path(ctx.value["path"]).expanduser().absolute()


def _sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with open(path, "rb") as handle:
        for chunk in iter(lambda: handle.read(_HASH_CHUNK), b""):
            digest.update(chunk)
    return digest.hexdigest()

File: next-go/internal/plugins/fileservice/fileservice.go
// Package fileservice implements unrestricted local path metadata and reads.
package fileservice

import (
	"context"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"unicode/utf8"

	"viewer/internal/busclient"
	"viewer/internal/plugins/pluginrpc"
)

const DefaultMaxReadBytes int64 = 1024 * 1024

var Manifest = busclient.Manifest{
	ID: "file-service", Version: "0.1.0",
	Slots: map[string]any{"resolve": map[string]any{}, "read": map[string]any{}, "hash": map[string]any{}},
	Emits: map[string]any{},
}

type Plugin struct{ client *busclient.Client }

func New() *Plugin { return &Plugin{} }

func (p *Plugin) Start(ctx context.Context, kernelWS string, managed bool) error {
	client := busclient.New(kernelWS, Manifest, busclient.WithManaged(managed))
	for pattern, handler := range map[string]func(busclient.Frame){
		"file:_:resolve": p.resolve,
		"file:_:read":    p.read,
		"file:_:hash":    p.hash,
	} {
		if _, err := client.Subscribe(pattern, handler); err != nil {
			_ = client.Close()
			return err
		}
	}
	p.client = client
	if err := client.Connect(ctx); err != nil {
		p.client = nil
		_ = client.Close()
		return err
	}
	slog.Info("file-service started")
	return nil
}

func (p *Plugin) Close() error {
	if p.client == nil {
		return nil
	}
	return p.client.Close()
}

func (p *Plugin) resolve(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	path, ok := requestPath(frame)
	if !ok {
		p.replyError(frame, "invalid_request", "missing required field: path")
		return
	}
	info, err := os.Stat(path)
	if os.IsNotExist(err) {
		p.replyError(frame, "not_found", "no such file: "+path)
		return
	}
	if err != nil {
		p.replyError(frame, "read_error", err.Error())
		return
	}
	var digest any
	if info.Mode().IsRegular() {
		hash, hashErr := sha256File(path)
		if hashErr != nil {
			p.replyError(frame, "read_error", hashErr.Error())
			return
		}
		digest = hash
	}
	p.reply(frame, map[string]any{
		"path": path, "exists": true, "is_dir": info.IsDir(),
		"size": info.Size(), "mtime": info.ModTime().Unix(), "sha256": digest,
	})
}

func (p *Plugin) read(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	path, ok := requestPath(frame)
	if !ok {
		p.replyError(frame, "invalid_request", "missing required field: path")
		return
	}
	value, _ := pluginrpc.Object(frame)
	maxBytes, validLimit := maxReadBytes(value)
	if !validLimit {
		p.replyError(frame, "invalid_request", "max_bytes must be an integer")
		return
	}
	info, err := os.Stat(path)
	if os.IsNotExist(err) {
		p.replyError(frame, "not_found", "no such file: "+path)
		return
	}
	if err != nil {
		p.replyError(frame, "read_error", err.Error())
		return
	}
	if info.Size() > maxBytes {
		p.replyError(frame, "too_large", fmt.Sprintf(
			"%s is %d bytes, above the %d-byte inline cap; use the gateway by-reference data plane",
			path, info.Size(), maxBytes,
		))
		return
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		p.replyError(frame, "read_error", err.Error())
		return
	}
	encoding := "utf-8"
	content := string(raw)
	if !utf8.Valid(raw) {
		encoding = "base64"
		content = base64.StdEncoding.EncodeToString(raw)
	}
	p.reply(frame, map[string]any{
		"path": path, "size": info.Size(), "encoding": encoding, "content": content,
	})
}

func (p *Plugin) hash(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	path, ok := requestPath(frame)
	if !ok {
		p.replyError(frame, "invalid_request", "missing required field: path")
		return
	}
	info, err := os.Stat(path)
	if os.IsNotExist(err) || err == nil && !info.Mode().IsRegular() {
		p.replyError(frame, "not_found", "no such file: "+path)
		return
	}
	if err != nil {
		p.replyError(frame, "read_error", err.Error())
		return
	}
	digest, err := sha256File(path)
	if err != nil {
		p.replyError(frame, "read_error", err.Error())
		return
	}
	p.reply(frame, map[string]any{"path": path, "sha256": digest})
}

func requestPath(frame busclient.Frame) (string, bool) {
	value, ok := pluginrpc.Object(frame)
	if !ok {
		return "", false
	}
	raw, ok := value["path"].(string)
	if !ok || raw == "" {
		return "", false
	}
	expanded, err := expandUser(raw)
	if err != nil {
		return "", false
	}
	absolute, err := filepath.Abs(expanded)
	if err != nil {
		return "", false
	}
	return absolute, true
}

func expandUser(path string) (string, error) {
	if path != "~" && !strings.HasPrefix(path, "~/") {
		return path, nil
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	if path == "~" {
		return home, nil
	}
	return filepath.Join(home, strings.TrimPrefix(path, "~/")), nil
}

func maxReadBytes(value map[string]any) (int64, bool) {
	raw, exists := value["max_bytes"]
	if !exists {
		return DefaultMaxReadBytes, true
	}
	switch number := raw.(type) {
	case float64:
		if number != float64(int64(number)) {
			return 0, false
		}
		return int64(number), true
	case string:
		parsed, err := strconv.ParseInt(number, 10, 64)
		return parsed, err == nil
	default:
		return 0, false
	}
}

func sha256File(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()
	digest := sha256.New()
	buffer := make([]byte, 64*1024)
	for {
		count, readErr := file.Read(buffer)
		if count > 0 {
			_, _ = digest.Write(buffer[:count])
		}
		if readErr != nil {
			if readErr == io.EOF {
				break
			}
			return "", readErr
		}
	}
	return fmt.Sprintf("%x", digest.Sum(nil)), nil
}

func (p *Plugin) reply(frame busclient.Frame, result any) {
	if err := pluginrpc.Respond(p.client, frame, result); err != nil {
		slog.Error("file-service RPC response failed", "error", err)
	}
}

func (p *Plugin) replyError(frame busclient.Frame, code, message string) {
	if err := pluginrpc.RespondError(p.client, frame, code, message); err != nil {
		slog.Error("file-service RPC error response failed", "error", err)
	}
}
